src/viztracer/modules/snaptrace.c

```diff
@@ -603,6 +613,8 @@ snaptrace_load(TracerObject* self, PyObject* args)
     PyObject* task_dict = NULL;
     PyObject* func_name_dict = PyDict_New();
 
+    SNAPTRACE_THREAD_PROTECT_START(self);
```

Comment:
This line should probably be moved to the beginning of the code? As on line 600, ``self->buffer`` is used to initialize ``curr``.

You can move ``curr`` initialization inside the protection as well(to save some time for the unicode string generation?). 

If you prefer declaration before the protection, initialize ``curr`` to ``null`` then assign it inside the protection would also be okay with me.

Reply:
Ooops. I missed that. I'll move it up earlier.

Reply:
I moved it up to immediately after the definition of `lst`. The `lst` declaration needs to be outside because `Py_BEGIN_CRITICAL_SECTION`/`Py_END_CRITICAL_SECTION` currently delimit a C scope and `lst` is returned at the end of the function. (Not clear if delimiting a C scope was a good idea or not...)

From a correctness point of view, it doesn't really matter if most of the definitions and initializations are inside or outside the protection. (Other than the initialization of `curr` as you pointed out).

I also don't think it matters from a performance point of view: I don't expect any contention when calling snaptrace_load/dump. (The only contention I expect is during tracing when calling `get_next_event`.)

Reply:
Huh, I did not realize that macro will introduce a C scope. Could you comment that under `#ifdef Py_NOGIL`? That is an unobvious thing and could potentially cause some trouble if the person using it is not aware. Just introduce a bit about the mechanism and the things to know when using them.

Other than that, I think the code is good to me.

Reply:
I added a comment before the macro definitions